### PR TITLE
 Implement Sink.takeLast operator

### PR DESCRIPTION
**Pull request by naferx**
Ref https://github.com/akka/akka/issues/25322

### CODE REVIEW

File: akka-docs/src/main/paradox/stream/operators/Sink/takeLast.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# Sink.takeLast
+
+Materializes into a @scala[`Future`] @java[`CompletionStage`] of  @scala[`Seq[T]`] @java[`List<In>`] containing the last `n` collected elements when the stream completes.
```
Comment (ktoso) [inline akka-docs/src/main/paradox/stream/operators/Sink/takeLast.md]:
```
@scala[`immutable.Seq[T]`]
```

File: akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala [hunk 2]
```diff
@@ -202,6 +202,13 @@ object Sink {
    */
   def lastOption[T]: Sink[T, Future[Option[T]]] = Sink.fromGraph(new LastOptionStage[T]).withAttributes(DefaultAttributes.lastOptionSink)
 
+  /**
+   * A `Sink` that materializes into a a `Future` of `Seq[T]` containing the last `n` collected elements.
+   * If the stream completes before signaling at least n elements, the Future will complete with the number
+   * of elements taken at that point.
+   */
+  def takeLast[T](n: Int): Sink[T, Future[immutable.Seq[T]]] = Sink.fromGraph(new TakeLastStage[T](n)).withAttributes(DefaultAttributes.takeLastSink)
```
[the remark this answers is not in the document]
Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala]: Yeah, let's do that for now, a simple if here is okey.
We could move all over to the seq one but let's not for now I think :)

**Review by ktoso — APPROVED** [review]
Small changes requested but looks great :)

File: akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala [hunk 3]
```diff
@@ -198,10 +198,23 @@ object Sink {
...
+  /**
+   * A `Sink` that materializes into a a `Future` of `immutable.Seq[T]` containing the last `n` collected elements.
+   * If the stream completes before signaling at least n elements, the Future will complete with the number
+   * of elements taken at that point.
+   * If the stream never completes, the `Future` will never complete.
+   * If there is a failure signaled in the stream the `Future` will be completed with failure.
+   */
+  def takeLast[T](n: Int): Sink[T, Future[immutable.Seq[T]]] =
+    if (n == 1)
+      lastOption.mapMaterializedValue(fut ⇒ fut.map(_.fold(immutable.Seq.empty[T])(m ⇒ immutable.Seq(m)))(ExecutionContexts.sameThreadExecutionContext))
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala]: `lastOption.named("takeLast")` perhaps?

**Review by ktoso — APPROVED** [review]
LGTM from me except one additional thing

**Review by jrudolph — COMMENTED** [review]
I'm fine with adding the API, but I don't see the reason why this shouldn't be implemented in terms of fold (same for last/lastOption/seq). As long as a significant performance benefit hasn't been proven, we shouldn't introduce new GraphStage implementations which we know to have a non-zero maintenance overhead (which is what we tell everyone else as well).

File: akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala [hunk 4]
```diff
@@ -157,12 +157,24 @@ object Sink {
...
-   * See also [[head]].
+   * See also [[head]], [[takeLast]].
    */
   def lastOption[In](): Sink[In, CompletionStage[Optional[In]]] =
     new Sink(scaladsl.Sink.lastOption[In].mapMaterializedValue(
       _.map(_.asJava)(ExecutionContexts.sameThreadExecutionContext).toJava))
 
+  /**
+   * A `Sink` that materializes into a a `CompletionStage` of `List<In>` containing the last `n` collected elements.
+   * If the stream completes before signaling at least n elements, the CompletionStage will complete with the number
```
[the remark this answers is not in the document]
Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala]: Good thanks

**Comment by ktoso** [comment]
I've implemented all of them using the Seq stage including the option one so we don't have two stages. 
I'll close this PR and merge the one with the small change, keeping authorship of @naferx ofc.

**Pull request closed without merging**